File: backend/app/schemas/manufacturing.py
from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, Field


class ManufacturingStatus(Enum):
    """
    Enum for manufacturing status.
    """

    QUEUED = "queued"
    IN_PROGRESS = "in_progress"
    CANCELLED = "cancelled"
    COMPLETED = "completed"


class ManufacturingBase(BaseModel):
    """
    Shared properties for manufacturing models. Visible by anyone.
    """

    client_uuid: UUID = Field(
        ..., description="The client associated with the manufacturing item."
    )
    project_uuid: UUID = Field(
        ..., description="The project associated with the manufacturing item."
    )
    ppo_id: str = Field(..., description="The PPO ID of the manufacturing item.")
    job_no: str = Field(..., description="The job number of the manufacturing item.")
    rfy_file: str = Field(..., description="The RFY file of the manufacturing item.")
    material_uuid: UUID = Field(
        ..., description="The material associated with the manufacturing item."
    )
    status: ManufacturingStatus = Field(
        ..., description="The status of the manufacturing item."
    )


class ManufacturingCreate(ManufacturingBase):
    """
    Properties to create a new manufacturing item. Visible by anyone.
    """

    pass


class ManufacturingUpdate(BaseModel):
    """
    Properties to update a manufacturing item. Visible by anyone.
    """

    client_uuid: UUID | None = Field(
        default=None, description="The client associated with the manufacturing item."
    )
    project_uuid: UUID | None = Field(
        default=None, description="The project associated with the manufacturing item."
    )
    ppo_id: str | None = Field(
        default=None, description="The PPO ID of the manufacturing item."
    )
    job_no: str | None = Field(
        default=None, description="The job number of the manufacturing item."
    )
    rfy_file: str | None = Field(
        default=None, description="The RFY file of the manufacturing item."
    )
    material_uuid: UUID | None = Field(
        default=None, description="The material associated with the manufacturing item."
    )
    status: ManufacturingStatus | None = Field(
        default=None, description="The status of the manufacturing item."
    )


class Manufacturing(ManufacturingBase):
    """
    Properties shared by models stored in the database. Visible by anyone.
    """

    uuid: UUID = Field(..., description="Unique identifier for the manufacturing item.")
    last_updated: datetime = Field(
        ..., description="Last updated timestamp for the manufacturing item."
    )

File: frontend/src/models/manufacturing.ts
import { Form } from './form'
import { Column } from './column'

export interface Manufacturing {
  uuid: string
  client_uuid: string
  project_uuid: string
  ppo_id: string
  job_no: string
  rfy_file: string
  material_uuid: string
  status: string
  last_updated: string
}

export const ManufacturingForm: Form = {
  title: 'Manufacturing Information',
  groups: [
    {
      title: 'Manufacturing Details',
      fields: [
        { name: 'client_uuid', label: 'Client UUID', type: 'text', required: true },
        { name: 'project_uuid', label: 'Project UUID', type: 'text', required: true },
        { name: 'ppo_id', label: 'PPO ID', type: 'text', required: true },
        { name: 'job_no', label: 'Job No', type: 'text', required: true },
        { name: 'rfy_file', label: 'RFY File', type: 'text', required: true },
        { name: 'material_uuid', label: 'Material UUID', type: 'text', required: true },
        {
          name: 'status',
          label: 'Status',
          type: 'select',
          required: true,
          options: [
            { label: 'Queued', value: 'queued' },
            { label: 'In Progress', value: 'in_progress' },
            { label: 'Cancelled', value: 'cancelled' },
            { label: 'Completed', value: 'completed' },
          ],
        },
      ],
    },
  ],
}

export const ManufacturingColumns: Column<Manufacturing>[] = [
  { key: 'client_uuid' as keyof Manufacturing, label: 'Client UUID' },
  { key: 'project_uuid' as keyof Manufacturing, label: 'Project UUID' },
  { key: 'ppo_id' as keyof Manufacturing, label: 'PPO ID' },
  { key: 'rfy_file' as keyof Manufacturing, label: 'RFY File' },
  { key: 'status' as keyof Manufacturing, label: 'Status' },
  { key: 'last_updated' as keyof Manufacturing, label: 'Last Updated' },
]
